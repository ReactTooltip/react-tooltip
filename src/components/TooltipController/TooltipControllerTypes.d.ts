import type { CSSProperties } from 'react'

import type {
  PlacesType,
  VariantType,
  WrapperType,
  ChildrenType,
  EventsType,
  PositionStrategy,
  IPosition,
} from 'components/Tooltip/TooltipTypes'

export interface ITooltipController {
  className?: string
  classNameArrow?: string
  content?: string
  html?: string
  place?: PlacesType
  offset?: number
  id?: string
  variant?: VariantType
  anchorId?: string
  wrapper?: WrapperType
  children?: ChildrenType
  events?: EventsType[]
  positionStrategy?: PositionStrategy
  delayShow?: number
  delayHide?: number
  float?: boolean
  noArrow?: boolean
  clickable?: boolean
  style?: CSSProperties
  position?: IPosition
  isOpen?: boolean
  setIsOpen?: (value: boolean) => void
}

declare module 'react' {
  interface HTMLAttributes<T> extends AriaAttributes, DOMAttributes<T> {
    'data-tooltip-place'?: PlacesType
    'data-tooltip-content'?: string
    'data-tooltip-html'?: string
    'data-tooltip-variant'?: VariantType
    'data-tooltip-offset'?: number
    'data-tooltip-wrapper'?: WrapperType
    'data-tooltip-events'?: EventsType[]
    'data-tooltip-position-strategy'?: PositionStrategy
    'data-tooltip-delay-show'?: number
    'data-tooltip-delay-hide'?: number
    'data-tooltip-float'?: boolean
  }
}
