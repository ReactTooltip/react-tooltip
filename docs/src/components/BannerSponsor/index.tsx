/* eslint-disable import/no-unresolved */
import React from 'react'
// @ts-ignore
import LogoFrigade from '@site/static/img/sponsors/frigade.png'
// @ts-ignore
import LogoDopt from '@site/static/img/sponsors/dopt.png'
import './styles.css'

declare global {
  interface Window {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    dataLayer?: any
  }
}

const SPONSORS = {
  frigade: {
    logo: LogoFrigade,
    title: 'Frigade',
  },
  dopt: {
    logo: LogoDopt,
    title: 'Dopt',
  },
}

interface BannerSponsorProps {
  sponsorKey: keyof typeof SPONSORS
}

const BannerSponsor = ({ sponsorKey }: BannerSponsorProps) => {
  const sponsor = SPONSORS[sponsorKey]

  const onClickSponsorBannerEventHandler = () => {
    if (typeof window !== 'undefined') {
      window.dataLayer = window.dataLayer || []

      window.dataLayer.push({
        event: `click_${sponsorKey}_banner`,
        place: 'sidebar',
      })
    }

    return true
  }

  return (
    <div className="sponsor-banner">
      <a
        href="https://frigade.com/?source=react-tooltip"
        title={sponsor.title}
        target="_blank"
        rel="noreferrer"
        onClick={onClickSponsorBannerEventHandler}
      >
        <img src={sponsor.logo} alt={sponsor.title} />
      </a>
    </div>
  )
}

export default BannerSponsor
