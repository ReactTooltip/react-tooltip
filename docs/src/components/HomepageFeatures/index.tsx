/* eslint-disable import/no-unresolved */
/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable global-require */
import React from 'react'
import clsx from 'clsx'
import styles from './styles.module.css'

type FeatureItem = {
  title: string
  Svg: React.ComponentType<React.ComponentProps<'svg'>>
  description: JSX.Element
}

const FeatureList: FeatureItem[] = [
  {
    title: 'Easy to Use',
    Svg: require('@site/static/img/only-tooltip-top.svg').default,
    description: (
      <>ReactTooltip was designed with love from the ground up to be easily installed and used.</>
    ),
  },
  {
    title: 'Open Source',
    Svg: require('@site/static/img/github.svg').default,
    description: (
      <>
        An Open Source project built by developers to developers, we work together with the
        community to always try to improve ReactTooltip.
      </>
    ),
  },
  {
    title: 'Focus on What Matters',
    Svg: require('@site/static/img/happy-face-tooltip.svg').default,
    description: (
      <>
        Don&rsquo;t waste your time building a tooltip component from scratch, just use ReactTooltip
        and focus on new features or bug fixes.
      </>
    ),
  },
]

function Feature({ title, Svg, description }: FeatureItem) {
  let classVar = styles.featureSvg;
  if(Svg === require('@site/static/img/github.svg').default)
  {
    classVar = styles.featureSvg +" "+ styles.githubLogo;
  }
  return (
    <div className={clsx('col col--4')}>
      <div className="text--center">
        <Svg className={classVar} role="img" />
      </div>
      <div className="text--center padding-horiz--md">
        <h3>{title}</h3>
        <p>{description}</p>
      </div>
    </div>
  )
}

export default function HomepageFeatures(): JSX.Element {
  return (
    <section className={styles.features}>
      <div className="container">
        <div className="row">
          {FeatureList.map((props, idx) => (
            // eslint-disable-next-line react/no-array-index-key
            <Feature key={idx} {...props} />
          ))}
        </div>
      </div>
    </section>
  )
}
