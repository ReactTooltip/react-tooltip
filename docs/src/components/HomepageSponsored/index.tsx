/* eslint-disable react/require-default-props */
/* eslint-disable import/no-unresolved */
/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable global-require */
import React from 'react'
import clsx from 'clsx'
import styles from './styles.module.css'

type FeatureItem = {
  title: string
  Svg?: React.ComponentType<React.ComponentProps<'svg'>>
  // eslint-disable-next-line @typescript-eslint/no-explicit-any, react/no-unused-prop-types
  src?: any
  // eslint-disable-next-line react/no-unused-prop-types
  eventTitle?: string
  link: string
}

const FeatureList: FeatureItem[] = [
  {
    title: 'Digital Ocean',
    Svg: require('@site/static/img/digital-ocean-powered-by.svg').default,
    link: 'https://www.digitalocean.com/?refcode=0813b3be1161&utm_campaign=Referral_Invite&utm_medium=Referral_Program&utm_source=badge',
  },
  {
    title: 'Algolia',
    Svg: require('@site/static/img/Algolia-logo.svg').default,
    link: 'https://docsearch.algolia.com/',
  },
]

const SponsorList: FeatureItem[] = [
  {
    title: 'Frigade',
    src: require('@site/static/img/sponsors/frigade.png').default,
    link: 'https://frigade.com/?source=react-tooltip',
    eventTitle: 'frigade',
  },
  {
    title: 'Dopt',
    src: require('@site/static/img/sponsors/dopt.png').default,
    link: 'https://dopt.com/?source=react-tooltip',
    eventTitle: 'dopt',
  },
]

function Feature({ title, Svg, link }: FeatureItem) {
  return (
    <div className={clsx('col col--6')}>
      <div className="text--center">
        <a href={link} title={title} target="_blank" rel="noreferrer">
          <Svg className={styles.featureSvg} role="img" />
        </a>
      </div>
    </div>
  )
}

export default function HomepageSponsored(): JSX.Element {
  const onClickSponsorBannerEventHandler = (title: string) => {
    if (typeof window !== 'undefined') {
      window.dataLayer = window.dataLayer || []

      window.dataLayer.push({
        event: `click_${title}_banner`,
        place: 'home',
      })
    }

    return true
  }

  return (
    <section className={styles.features}>
      <div className="container">
        <h3 className={styles.sponsoredTitle}>Sponsored by</h3>
        <div className="row">
          {SponsorList.map(({ link, title, src, eventTitle }, idx) => (
            // eslint-disable-next-line react/no-array-index-key
            <div key={idx} className={clsx(`col col--${12 / SponsorList.length}`)}>
              <div className="text--center">
                <a
                  href={link}
                  title={title}
                  target="_blank"
                  rel="noreferrer"
                  onClick={() => {
                    onClickSponsorBannerEventHandler(eventTitle)
                  }}
                >
                  <img src={src} alt={title} width={480} />
                </a>
              </div>
            </div>
          ))}
        </div>
      </div>
      <div className="container">
        <h3 className={styles.sponsoredTitle}>Powered by</h3>
        <div className="row">
          {FeatureList.map((props, idx) => (
            // eslint-disable-next-line react/no-array-index-key
            <Feature key={idx} {...props} />
          ))}
        </div>
      </div>
    </section>
  )
}
