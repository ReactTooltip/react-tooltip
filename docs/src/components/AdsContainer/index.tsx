/* eslint-disable import/no-unresolved */
import React from 'react'
import AdsContainerElement from '@site/src/components/AdsContainerElement'
import BannerSponsor from '@site/src/components/BannerSponsor'
import './styles.css'

const AdsContainer = () => {
  return (
    <div className="ads-container fixed">
      <BannerSponsor sponsorKey="frigade" tier="gold" />
      <AdsContainerElement />
      <BannerSponsor sponsorKey="dopt" tier="silver" />
    </div>
  )
}

export default AdsContainer
