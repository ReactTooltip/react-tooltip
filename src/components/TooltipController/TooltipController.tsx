import { useEffect, useState } from 'react'
import { Tooltip } from 'components/Tooltip'
import type {
  EventsType,
  PositionStrategy,
  PlacesType,
  VariantType,
  WrapperType,
  DataAttribute,
  ITooltip,
} from 'components/Tooltip/TooltipTypes'
import { useTooltip } from 'components/TooltipProvider'
import type { ITooltipController } from './TooltipControllerTypes'

const TooltipController = ({
  id,
  anchorId,
  content,
  html,
  className,
  classNameArrow,
  variant = 'dark',
  place = 'top',
  offset = 10,
  wrapper = 'div',
  children = null,
  events = ['hover'],
  positionStrategy = 'absolute',
  delayShow = 0,
  delayHide = 0,
  float = false,
  noArrow = false,
  clickable = false,
  style,
  position,
  isOpen,
  setIsOpen,
}: ITooltipController) => {
  const [tooltipContent, setTooltipContent] = useState(content || html)
  const [tooltipPlace, setTooltipPlace] = useState(place)
  const [tooltipVariant, setTooltipVariant] = useState(variant)
  const [tooltipOffset, setTooltipOffset] = useState(offset)
  const [tooltipDelayShow, setTooltipDelayShow] = useState(delayShow)
  const [tooltipDelayHide, setTooltipDelayHide] = useState(delayHide)
  const [tooltipFloat, setTooltipFloat] = useState(float)
  const [tooltipWrapper, setTooltipWrapper] = useState<WrapperType>(wrapper)
  const [tooltipEvents, setTooltipEvents] = useState(events)
  const [tooltipPositionStrategy, setTooltipPositionStrategy] = useState(positionStrategy)
  const [isHtmlContent, setIsHtmlContent] = useState(Boolean(html))
  const { anchorRefs, activeAnchor } = useTooltip()(id)

  const getDataAttributesFromAnchorElement = (elementReference: HTMLElement) => {
    const dataAttributes = elementReference?.getAttributeNames().reduce((acc, name) => {
      if (name.startsWith('data-tooltip-')) {
        const parsedAttribute = name.replace(/^data-tooltip-/, '') as DataAttribute
        acc[parsedAttribute] = elementReference?.getAttribute(name) ?? null
      }
      return acc
    }, {} as Record<DataAttribute, string | null>)

    return dataAttributes
  }

  const applyAllDataAttributesFromAnchorElement = (
    dataAttributes: Record<string, string | null>,
  ) => {
    const handleDataAttributes: Record<DataAttribute, (value: string | null) => void> = {
      place: (value) => {
        setTooltipPlace((value as PlacesType) ?? place)
      },
      content: (value) => {
        setTooltipContent(value ?? content)
      },
      html: (value) => {
        setIsHtmlContent(Boolean(value ?? html))
        setTooltipContent(value ?? html ?? content)
      },
      variant: (value) => {
        setTooltipVariant((value as VariantType) ?? variant)
      },
      offset: (value) => {
        setTooltipOffset(value === null ? offset : Number(value))
      },
      wrapper: (value) => {
        setTooltipWrapper((value as WrapperType) ?? 'div')
      },
      events: (value) => {
        const parsed = value?.split(' ') as EventsType[]
        setTooltipEvents(parsed ?? events)
      },
      'position-strategy': (value) => {
        setTooltipPositionStrategy((value as PositionStrategy) ?? positionStrategy)
      },
      'delay-show': (value) => {
        setTooltipDelayShow(value === null ? delayShow : Number(value))
      },
      'delay-hide': (value) => {
        setTooltipDelayHide(value === null ? delayHide : Number(value))
      },
      float: (value) => {
        setTooltipFloat(value === null ? float : Boolean(value))
      },
    }
    // reset unset data attributes to default values
    // without this, data attributes from the last active anchor will still be used
    Object.values(handleDataAttributes).forEach((handler) => handler(null))
    Object.entries(dataAttributes).forEach(([key, value]) => {
      handleDataAttributes[key as DataAttribute]?.(value)
    })
  }

  useEffect(() => {
    setIsHtmlContent(false)
    setTooltipContent(content)
    if (html) {
      // html will take precedence
      setIsHtmlContent(true)
      setTooltipContent(html)
    }
  }, [content, html])

  useEffect(() => {
    const elementRefs = new Set(anchorRefs)

    const anchorById = document.querySelector(`[id='${anchorId}']`) as HTMLElement
    if (anchorById) {
      elementRefs.add({ current: anchorById })
    }

    if (!elementRefs.size) {
      return () => null
    }

    const observerCallback: MutationCallback = (mutationList) => {
      mutationList.forEach((mutation) => {
        if (
          !activeAnchor.current ||
          mutation.type !== 'attributes' ||
          !mutation.attributeName?.startsWith('data-tooltip-')
        ) {
          return
        }
        // make sure to get all set attributes, since all unset attributes are reset
        const dataAttributes = getDataAttributesFromAnchorElement(activeAnchor.current)
        applyAllDataAttributesFromAnchorElement(dataAttributes)
      })
    }

    // Create an observer instance linked to the callback function
    const observer = new MutationObserver(observerCallback)

    // do not check for subtree and childrens, we only want to know attribute changes
    // to stay watching `data-attributes-*` from anchor element
    const observerConfig = { attributes: true, childList: false, subtree: false }

    const element = activeAnchor.current ?? anchorById

    if (element) {
      const dataAttributes = getDataAttributesFromAnchorElement(element)
      applyAllDataAttributesFromAnchorElement(dataAttributes)
      // Start observing the target node for configured mutations
      observer.observe(element, observerConfig)
    }

    return () => {
      // Remove the observer when the tooltip is destroyed
      observer.disconnect()
    }
  }, [anchorRefs, activeAnchor, anchorId])

  const props: ITooltip = {
    id,
    anchorId,
    className,
    classNameArrow,
    content: tooltipContent,
    isHtmlContent,
    place: tooltipPlace,
    variant: tooltipVariant,
    offset: tooltipOffset,
    wrapper: tooltipWrapper,
    events: tooltipEvents,
    positionStrategy: tooltipPositionStrategy,
    delayShow: tooltipDelayShow,
    delayHide: tooltipDelayHide,
    float: tooltipFloat,
    noArrow,
    clickable,
    style,
    position,
    isOpen,
    setIsOpen,
  }

  return children ? <Tooltip {...props}>{children}</Tooltip> : <Tooltip {...props} />
}

export default TooltipController
