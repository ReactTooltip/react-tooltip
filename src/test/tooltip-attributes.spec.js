import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import '@testing-library/jest-dom'
import { TooltipController as Tooltip } from '../components/TooltipController'

// Tell Jest to mock all timeout functions
jest.useRealTimers()

// eslint-disable-next-line react/prop-types
const TooltipAttrs = ({ id, ...anchorParams }) => (
  <>
    <span id={id} {...anchorParams}>
      Lorem Ipsum
    </span>
    <Tooltip anchorId={id} />
  </>
)

describe('tooltip attributes', () => {
  test('tooltip without element reference', async () => {
    const { container } = render(<TooltipAttrs data-tooltip-content="Hello World!" />)
    const anchorElement = screen.getByText('Lorem Ipsum')

    await userEvent.hover(anchorElement)

    await waitFor(() => {
      expect(screen.queryByText('Hello World!')).not.toBeInTheDocument()
      expect(screen.queryByRole('tooltip')).not.toBeInTheDocument()
    })

    expect(container).toMatchSnapshot()
  })

  test('basic tooltip', async () => {
    const { container } = render(
      <TooltipAttrs id="basic-example-attr" data-tooltip-content="Hello World!" />,
    )
    const anchorElement = screen.getByText('Lorem Ipsum')

    await userEvent.hover(anchorElement)

    let tooltip = null

    await waitFor(() => {
      tooltip = screen.getByRole('tooltip')
    })

    expect(anchorElement).toHaveAttribute('data-tooltip-content')
    expect(tooltip).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  test('tooltip with place', async () => {
    const { container } = render(
      <TooltipAttrs
        id="example-place-attr"
        data-tooltip-content="Hello World!"
        data-tooltip-place="right"
      />,
    )
    const anchorElement = screen.getByText('Lorem Ipsum')

    await userEvent.hover(anchorElement)

    let tooltip = null

    await waitFor(() => {
      tooltip = screen.getByRole('tooltip')
    })

    expect(anchorElement).toHaveAttribute('data-tooltip-place')
    expect(anchorElement).toHaveAttribute('data-tooltip-content')
    expect(tooltip).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })
})
