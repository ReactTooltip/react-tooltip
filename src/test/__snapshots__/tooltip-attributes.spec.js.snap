// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tooltip attributes basic tooltip 1`] = `
<div>
  <span
    data-tooltip-content="Hello World!"
    id="basic-example-attr"
  >
    Lorem Ipsum
  </span>
  <div
    class="react-tooltip"
    role="tooltip"
  >
    Hello World!
    <div
      class="react-tooltip-arrow"
    />
  </div>
</div>
`;

exports[`tooltip attributes tooltip with place 1`] = `
<div>
  <span
    data-tooltip-content="Hello World!"
    data-tooltip-place="right"
    id="example-place-attr"
  >
    Lorem Ipsum
  </span>
  <div
    class="react-tooltip"
    role="tooltip"
  >
    Hello World!
    <div
      class="react-tooltip-arrow"
    />
  </div>
</div>
`;

exports[`tooltip attributes tooltip without element reference 1`] = `
<div>
  <span
    data-tooltip-content="Hello World!"
  >
    Lorem Ipsum
  </span>
</div>
`;
